lib/provisioning/ProvisionDispatcher.js

```diff
@@ -0,0 +1,355 @@
+            }
+        });
+        // now write data in owners
+        return Object.keys(provisionsByOwner).forEach(owner => {
+            const zkPath = `${ProvisionDispatcher.constants.OWNERS}/${owner}`;
+            this.log.debug('set new provisions to owner',
+                           { zkPath: this.zkEndpoint + zkPath,
+                             provisionList: provisionsByOwner[owner] });
+            this.client.setData(
+                zkPath,
+                new Buffer(JSON.stringify(provisionsByOwner[owner])),
+                -1, // version
+                err => {
+                    if (err &&
+                        err.getCode() !== zookeeper.Exception.NO_NODE) {
+                        this.log.error('error in setData',
+                                       { zkPath: this.zkEndpoint + zkPath,
+                                         error: err });
+                    }
+                });
+        });
+    }
+
+    _registerLeader(cb) {
+        // register in election queue
+        const zkPath = (ProvisionDispatcher.constants.LEADERS +
+                        ProvisionDispatcher.constants.LEADER);
+        this.client.create(
+            zkPath, null,
+            zookeeper.ACL.OPEN_ACL_UNSAFE,
+            zookeeper.CreateMode.EPHEMERAL_SEQUENTIAL,
+            (err, _path) => {
+                if (err) {
+                    return cb(err);
+                }
+                this.myLeaderName = path.basename(_path);
+                this.log.debug('registered leader zk node',
+                               { zkPath: this.zkEndpoint + _path });
+                return cb();
+            });
+    }
+
+    _unregisterLeader(cb) {
+        // de-register from election queue
+        const zkPath = `${ProvisionDispatcher.constants.LEADERS}/` +
+                  `${this.myLeaderName}`;
+        this.client.remove(zkPath, -1, err => {
+            if (err) {
+                return cb(err);
+            }
+            this.log.debug('unregistered leader zk node',
+                           { zkPath: this.zkEndpoint + zkPath });
+            return cb();
+        });
+    }
+
+    _registerOwner(cb) {
+        this.myName = _genRandomHexChars(12);
+        const zkPath = `${ProvisionDispatcher.constants.OWNERS}/` +
+                  `${this.myName}`;
+        this.client.create(
+            zkPath, null, zookeeper.ACL.OPEN_ACL_UNSAFE,
+            zookeeper.CreateMode.EPHEMERAL,
+            err => {
+                if (err) {
+                    return cb(err);
+                }
+                this.log.debug('registered owner zk node',
+                               { zkPath: this.zkEndpoint + zkPath });
+                return cb(null, zkPath);
+            });
+    }
+
+    _unregisterOwner(cb) {
+        const zkPath = `${ProvisionDispatcher.constants.OWNERS}/` +
+                  `${this.myName}`;
+        this.client.remove(zkPath, -1, err => {
+            if (err) {
```

Comment:
I think it's ok to forward the error, and let the client handle it appropriately. Normally he should only call `unsubscribe()` if he already has called `subscribe()` earlier, in which case there shall be no error, or if he calls `unsubscribe()` in a "safety net" fashion without knowing if he subscribed earlier, he should then deliberately ignore the `NO_NODE` errors.